bench/src/bulk.rs

```diff
@@ -80,34 +136,128 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
+
+        duration_hist
+            .record(result.duration.as_millis() as u64)
+            .unwrap();
+        throughput_hist.record(result.throughput as u64).unwrap();
     }
-    stream.finish().await.context("failed finishing stream")?;
-    println!("sent {} bytes in {:?}", 1024 * DATA.len(), start.elapsed());
+
+    let dt = start.elapsed();
+    println!("Overall stats:\n");
+    println!(
+        "Sent {} bytes on {} streams in {:4.2?} ({:.2} MiB/s)\n",
+        total_size,
+        config.streams,
+        dt,
+        throughput_bps(dt, total_size as u64) / 1024.0 / 1024.0
+    );
+
+    println!("Stream metrics:\n");
+
+    let print_metric = |label: &'static str, get_metric: fn(&Histogram<u64>) -> u64| {
+        println!(
+            "{}: Throughput: {:7.2} MiB/s, Duration: {:>9}",
+            label,
+            get_metric(&throughput_hist) as f64 / 1024.0 / 1024.0,
+            format!("{:.2?}", Duration::from_millis(get_metric(&duration_hist)))
+        );
+    };
+
+    print_metric("AVG ", |hist| hist.mean() as u64);
+    print_metric("P0  ", |hist| hist.value_at_quantile(0.00));
+    print_metric("P10 ", |hist| hist.value_at_quantile(0.10));
+    print_metric("P50 ", |hist| hist.value_at_quantile(0.50));
+    print_metric("P90 ", |hist| hist.value_at_quantile(0.90));
+    print_metric("P100", |hist| hist.value_at_quantile(1.00));
+
+    // Explicit close of the connection, since handles can still be around due
+    // to `Arc`ing them
+    connection.close(0u32.into(), b"Benchmark done");
+
+    endpoint.wait_idle().await;
+
     Ok(())
 }
 
+fn throughput_bps(duration: Duration, size: u64) -> f64 {
+    (size as f64) / (duration.as_secs_f64())
+}
+
 fn rt() -> Runtime {
     Builder::new()
         .basic_scheduler()
         .enable_all()
         .build()
         .unwrap()
 }
+
+fn transport_config() -> quinn::TransportConfig {
+    // High stream windows are chosen because the amount of concurrent streams
+    // is configurable as a parameter.
+    let mut config = quinn::TransportConfig::default();
+    config.stream_window_uni(1024).unwrap();
+    config
+}
+
+#[derive(StructOpt, Debug)]
+#[structopt(name = "bulk")]
+struct Opt {
+    /// The total number of streams which should be created
+    #[structopt(long = "streams", short = "n", default_value = "1")]
+    streams: usize,
+    /// The amount of concurrent streams which should be used
+    #[structopt(long = "max_streams", short = "m", default_value = "1")]
+    max_streams: usize,
+    /// The amount of data to transfer on a stream in megabytes
+    #[structopt(long = "stream_size", default_value = "1024")]
+    stream_size_mb: usize,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct BenchConfig {
+    /// The total number of streams which should be created
+    streams: usize,
+    /// The amount of concurrent streams which should be used
+    max_streams: usize,
+    /// The amount of data to transfer on a stream in megabytes
+    stream_size_mb: usize,
+}
```

Comment:
Can we really not just derive `Clone + Copy` on `Opt`?

Reply:
Sure. I tried to separate the CLI options and function arguments. But for such a small program it likely doesn't matter.